src/tracks/playabletrack/wavetrack/ui/WaveClipTrimHandle.cpp

```diff
@@ -64,48 +71,105 @@ class WaveClipTrimHandle::AdjustBorder final : public WaveClipTrimHandle::ClipTr
       }
    }
 
+   //Search for a good snap points among all tracks, including
+   //the one to which adjusted clip belongs to, but not counting
+   //the borders of adjusted clip
+   static SnapPointArray FindSnapPoints(
+      const WaveTrack* currentTrack, 
+      WaveClip* adjustedClip,
+      const std::pair<double, double> range)
+   {
+      SnapPointArray result;
+
+      auto addSnapPoint = [&](double t, const Track* track)
+      {
+         if(t > range.second || t < range.first)
+            return;
+
+         for(const auto& snapPoint : result)
+            if(snapPoint.t == t)
+               return;
+         result.emplace_back(t, track);
+      };
+
+      if(const auto trackList = currentTrack->GetOwner())
+      {
+         for(const auto track : trackList->Any())
+         {
+            if(track == currentTrack ||
+               track->GetLinkType() == Track::LinkType::Aligned && *trackList->FindLeader(currentTrack) == track ||
+               currentTrack->GetLinkType() == Track::LinkType::Aligned && *trackList->FindLeader(track) == currentTrack)
+            {
+               //serching on the same track is a special case
+               for(const auto& clip : static_cast<const WaveTrack*>(track)->GetClips())
+               {
+                  //exclude adjusted clip
+                  if(clip.get() == adjustedClip)
+                     continue;
+
+                  addSnapPoint(clip->GetPlayStartTime(), track);
+                  addSnapPoint(clip->GetPlayEndTime(), track);
+               }
+            }
+            else
+            {
+               for (const auto &interval : track->GetIntervals()) 
+               {
+                  addSnapPoint(interval.Start(), track);
+                  if (interval.Start() != interval.End())
+                     addSnapPoint(interval.End(), track);
+               }
+            }
+         }
+      }
+      return result;
+   }
+
 public:
-   AdjustBorder(WaveTrack* track, std::shared_ptr<WaveClip>& clip, bool left)
-      : mAdjustingLeftBorder(left)
+   AdjustBorder(std::shared_ptr<WaveTrack>& track, std::shared_ptr<WaveClip>& clip, bool leftBorder, const ZoomInfo& zoomInfo)
+      : mTrack(track),
```

Comment:
The two shared_ptr arguments could be const references, and so they should, to aid the human reader.  Instead I read this then stopped to see whether `track` and `clip` were being reassigned for some reason.